kiali/kiali
# I18n support

### Describe the change

This PR adds i18n support to Kiali application using `i18next` library and some examples of Chinese translation in different kind of React components (classes, hooks, etc.). Translation of Kiali application in any language is out of scope of this PR.

Translation examples provided are:.
- Navigation menu (external function)
- Home cluster name tooltip in the top navigation menu (React hook)
- Number of labels per namespace (React class)

![image](https://github.com/kiali/kiali/assets/122779323/247caf40-ba25-4eaa-a59c-64ec6be9003d)

A language dropdown has been created to switch between English and Chinese. New locale value is stored in redux store so the value is persisted on refresh)

![image](https://github.com/kiali/kiali/assets/122779323/dffd7673-8351-4678-908d-ab0c6d46c24e)

For the moment supported languages are English (default) and Chinese (in progress).

Translations are stored in json files in `public/locales/${LOCALE}` folder and generated automatically by `i18next-parser` library on yarn command `yarn i18n`. This command is executed automatically on build process to add new statements to translation files (or removed unused ones). English translation file must not be modified since code text is the source of truth for English statements.

I have added two i18n settings in the Kiali server configuration under `kiali_feature_flags/ui_defaults/i18n`:
- language: Default language (if no value is stored in redux store)
- showSelector: If true shows selector language icon

Since Kiali code is the base of OSSMC plugin, I have tried to follow OpenShift dynamic plugin requirements for i18n labels (https://github.com/openshift/console-plugin-template/tree/main?tab=readme-ov-file#i18n):
- Added I18N_NAMESPACE variable as i18n translation namespace (empty for Kiali, proper plugin value for OSSMC).

```
const { t } = useTranslation(I18N_NAMESPACE);
```

- Key should be English value since key separators are not allowed.

```
t('Traffic Graph')
```

### How to translate different React components
**React class**: 
Use `withTranslation` high order component (https://react.i18next.com/latest/withtranslation-hoc)

You have to embed React class with `withTranslation` function (similar to redux `connect` function):
```
export const Menu = withTranslation(I18N_NAMESPACE)(MenuComponent);
```

Besides, `WithTranslation` props must be added to React class properties (like Redux properties)

```
type MenuProps = WithTranslation & {
   ...
};
```

**React hook**
Use `useTranslation` hook (https://react.i18next.com/latest/usetranslation-hook)

```
const { t } = useTranslation(I18N_NAMESPACE);
```

**External function**
Include `t` function as input parameter:

```
const navMenuItems = (t: TFunction): MenuItem[] => {
  return [
    {
      id: 'overview',
      title: t('Overview'),
      to: '/overview',
      pathsActive: [/^\/overview\/(.*)/]
    },


const allNavMenuItems = navMenuItems(this.props.t);
```

**Interpolation**
You can include any variable value within the translated statement with interpolation (https://www.i18next.com/translation-function/interpolation)

```
<div>{t('Kiali home cluster: {{name}}', { name: homeCluster?.name })}</div>
```

The variables are represented with `{{$variable}}`. You can set any name to the variable (in the example `name`), you have just to be sure that matches with json field. 

This is how it looks like in the translation file:
```
"Kiali home cluster: {{name}}": "Kiali主集群: {{name}}",
```

**Plurals**
When a statement is different in singular and plural, you can use defaultValue_one and defaultValue_other fields to add different plural statements in English (no need to add 's' manually anymore). Note that in this case variable name must be `count`. For convenience the key will be the singular statement. More info here: https://www.i18next.com/translation-function/plurals

```
labelsInfo = this.props.t('{{count}} label', {
        count: labelsLength,
        defaultValue_one: '{{count}} label',
        defaultValue_other: '{{count}} labels'
      });
```

**Formatting**
This feature is out of scope of this PR, although it would be interesting to investigate how to format dates for different languages.

### Steps to test the PR
1. Execute `yarn install` to install new i18n libraries
3. Check that switch language dropdown is available
4. Change the language to Chinese and see translated labels

### Automation testing
I have added unit test for `LanguageSwitch` component. For the moment I don't think Cypress test is needed since i18n language switch is disabled by default and Chinese translation is not done yet.

### Issue reference
Closes #2923 

Operator PR: https://github.com/kiali/kiali-operator/pull/746
